activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -54,16 +54,17 @@ def not(opts, *rest)
       end
     end
 
+    FROZEN_EMPTY_ARRAY = [].freeze
```

Comment:
There _is_ a need for freeze here; without it, it is possible to accidentally mutate this value. @sgrif argued against `freeze`ing the memoized `empty`s because their APIs do not expose mutating methods, but `Array`'s and `Hash`'s APIs do, and `freeze` here and below provides an important guarantee.